apps/sso/serializers.py

```diff
@@ -0,0 +1,120 @@
+from oauth2_provider.generators import generate_client_secret
+from oauth2_provider.models import (
+    get_access_token_model,
+    get_application_model,
+    get_grant_model,
+    get_refresh_token_model,
+)
+from rest_framework import serializers
+
+from apps.sso.models import ApplicationConsent
+
+# TODO:
+# Remove oauth2 from names, find a better naming
+
+
+class Oauth2ClientNonSensitiveSerializer(serializers.ModelSerializer):
+    client_type = serializers.ChoiceField(
+        choices=["public", "confidential"], allow_blank=True, default="public"
+    )
+    algorithm = serializers.HiddenField(default="")
+    user = serializers.HiddenField(default=serializers.CurrentUserDefault())
+
+    class Meta:
+        model = get_application_model()
+        fields = [
+            "id",
+            "client_id",
+            "client_type",
+            "authorization_grant_type",
+            "name",
+            "skip_authorization",
+            "created",
+            "updated",
+            "algorithm",
+            "scopes",
+            "website_url",
+            "terms_url",
+            "logo",
+            "contact_email",
+            "user",
+        ]
+
+    def create(self, validated_data):
+        client_secret = ""
+        if validated_data.get("client_type") == "confidential":
+            client_secret = generate_client_secret()
+        data = {
+            **validated_data,
+            "client_secret": client_secret,
+        }
+        client = get_application_model().objects.create(**data)
+        client.save()
+        return client
+
+    def update(self, instance, validated_data):
+        client_secret = instance.client_secret
+        had_oidc_enabled = instance.algorithm == "RS256"
+        will_enable_oidc = (
+            not had_oidc_enabled
+            and validated_data.get("scopes")
+            and "openid" in validated_data.get("scopes")
+        )
+        algorithm = "RS256" if had_oidc_enabled or will_enable_oidc else ""
+
+        if (
+            validated_data.get("client_type") == "confidential"
+            and not instance.client_secret
+        ):
+            client_secret = generate_client_secret()
+        elif validated_data.get("client_type") == "public" and instance.client_secret:
+            client_secret = ""
+        data = {
+            **validated_data,
+            "client_secret": client_secret,
+            "algorithm": algorithm,
+        }
+        for key, value in data.items():
+            setattr(instance, key, value)
+        instance.save()
+        return instance
+
+
+class Oauth2ClientSerializer(serializers.ModelSerializer):
+    redirect_uris = serializers.SerializerMethodField()
+
+    class Meta:
+        model = get_application_model()  # Allows for swappable application model
+        fields = "__all__"
```

Comment:
Hm, you're right! This serializer is only meant for endpoints that are capable of handling sensitive information, and at the moment the endpoint both requires a certain scope and that a user can only see their own clients' information. However, we might still want to have the full list instead so that it's obvious what kind of information we give out. And perhaps a name that indicates how "dangerous" it might be